libazureinit/src/media.rs

```diff
@@ -66,27 +68,94 @@ fn default_preprov_type() -> String {
-    let permissions = metadata.permissions();
-    let mut new_permissions = permissions.clone();
-    new_permissions.set_mode(0o700);
-    fs::set_permissions(file_path, new_permissions)?;
+pub struct Media<State = Unmounted> {
+    device_path: PathBuf,
+    mount_path: PathBuf,
+    state: std::marker::PhantomData<State>,
+}
+
+impl Media<Unmounted> {
+    pub fn new(device_path: PathBuf, mount_path: PathBuf) -> Media<Unmounted> {
+        Media {
+            device_path,
+            mount_path,
+            state: std::marker::PhantomData,
+        }
+    }
+
+    // The only thing you can do with the Media struct is call mount which, if it succeeds,
+    // moves it into the Mounted state. From there the only thing you can do with it is
+    // call unmount(), but you could make other functionality available by adding to the
+    // impl Media<Mounted> block.
+    pub fn mount(self) -> Result<Media<Mounted>, Error> {
+        create_dir_all(&self.mount_path)?;
+
+        let metadata = fs::metadata(&self.mount_path)?;
+        let permissions = metadata.permissions();
+        let mut new_permissions = permissions.clone();
+        new_permissions.set_mode(0o700);
+        fs::set_permissions(&self.mount_path, new_permissions)?;
+
+        let mount_status = Command::new("mount")
+            .arg("-o")
+            .arg("ro")
+            .arg(&self.device_path)
+            .arg(&self.mount_path)
+            .status()?;
 
-    Ok(())
+        if !mount_status.success() {
+            Err(Error::SubprocessFailed {
+                command: "mount".to_string(),
+                status: mount_status,
+            })
+        } else {
+            Ok(Media {
+                device_path: self.device_path,
+                mount_path: self.mount_path,
+                state: std::marker::PhantomData,
+            })
+        }
+    }
 }
 
-pub fn read_ovf_env_to_string() -> Result<String, Error> {
-    let file_path = "/run/azure-init/tmp/ovf-env.xml";
-    let mut file = File::open(file_path)?;
-    let mut contents = String::new();
-    file.read_to_string(&mut contents)?;
+impl Media<Mounted> {
+    pub fn unmount(self) -> Result<(), Error> {
+        let umount_status =
+            Command::new("umount").arg(self.mount_path).status()?;
+        if !umount_status.success() {
+            return Err(Error::SubprocessFailed {
+                command: "umount".to_string(),
+                status: umount_status,
+            });
+        }
+
+        let eject_status =
+            Command::new("eject").arg(self.device_path).status()?;
+        if !eject_status.success() {
+            Err(Error::SubprocessFailed {
+                command: "eject".to_string(),
+                status: eject_status,
+            })
+        } else {
+            Ok(())
+        }
+    }
 
-    Ok(contents)
+    pub fn read_ovf_env_to_string() -> Result<String, Error> {
+        let file_path = PATH_MOUNT_POINT.to_owned() + "ovf-env.xml";
```

Comment:
```suggestion
    pub fn read_ovf_env_to_string(&self) -> Result<String, Error> {
        let file_path = self.mount_point.clone().push("ovf-env.xml");
```

It would be surprising if you created the `Media` struct with a different mount path and then the "default" one got used. This also requires you to have a mounted `Media` instance.

Reply:
Right, thanks.